include/util/query_heap.hpp

```diff
@@ -289,7 +289,7 @@ class QueryHeap
         return inserted_nodes[index].node == node;
     }
 
-    boost::optional<HeapNode &> GetHeapNodeIfWasInserted(const NodeID node)
+    std::optional<HeapNode> GetHeapNodeIfWasInserted(NodeID node)
```

Comment:
I.e. now on each call we will copy HeapNode, right? How do we know that it won’t impact performance? May be worth refactoring it to `HeapNode*` instead(I mean use just pointer instead of optional<HeapNode> here)? From some perspective optional<T&> is just overcomplicated way to express simple pointer…

WDYT?


P.S. I didn’t have a time yet to look over the rest of PR, just something important what I noticed…

Reply:
More than it: here you can see an example when we change contents of node in a heap and that by reference… https://github.com/Project-OSRM/osrm-backend/blob/d51631401eeab6e90fa58b29f2ba8c0f56dc1161/include/engine/routing_algorithms/routing_base_mld.hpp#L331

I.e. if we return it not by reference now most likely existing logic is broken.

Reply:
I did run my build and it did at least not segfault or got killed by kernel for too much RAM.
To cite from cppreference.com:
"There are no optional references; a program is ill-formed if it instantiates an optional with a reference type."
Search engine would suggest std::reference_wrapper for this task.

Reply:
At the same time as I can see if we just change it to simple pointer instead of optional(I.e. won’t use neither boost::optional nor std::optional) the rest of code which uses this QueryHeap would “just work” without changes we had to make there…

Reply:
Regarding the other comment std::optional defines an operator->. However in some cases you probably would need to change a deref etc. I'm just testing solely using pointers for these parts. Does not work out of the box, but the idea should be doable.

Reply:
`boost::optional<T&>` seems to be effectively the same as `T*`, isn’t it?

Reply:
Yes it is. Still forking the main branch I need replace some of the idioms. Do you prefer nullptr or some void coercion ?

Reply:
What is void coercion? You mean `if (ptr == nullptr)` vs `if (!ptr)`? Personally I like first option as more explicit, but I don’t think it would be blocker to merge this PR :)

Reply:
Actually it's dreadfully simple, memory safety should be fine since it tests. I got too fed up with the nullptr version.
I think you need to approve CI again.